addons/binding/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/nhc2/NikoHomeControlCommunication2.java

```diff
@@ -0,0 +1,814 @@
+import java.util.stream.IntStream;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.io.transport.mqtt.MqttException;
+import org.eclipse.smarthome.io.transport.mqtt.MqttMessageSubscriber;
+import org.openhab.binding.nikohomecontrol.internal.protocol.NhcControllerEvent;
+import org.openhab.binding.nikohomecontrol.internal.protocol.NikoHomeControlCommunication;
+import org.openhab.binding.nikohomecontrol.internal.protocol.NikoHomeControlConstants.ActionType;
+import org.openhab.binding.nikohomecontrol.internal.protocol.nhc2.NhcDevice2.NhcProperty;
+import org.openhab.binding.nikohomecontrol.internal.protocol.nhc2.NhcMessage2.NhcMessageParam;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.FieldNamingPolicy;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonSyntaxException;
+import com.google.gson.reflect.TypeToken;
+
+/**
+ * The {@link NikoHomeControlCommunication2} class is able to do the following tasks with Niko Home Control II
+ * systems:
+ * <ul>
+ * <li>Start and stop MQTT connection with Niko Home Control II Connected Controller.
+ * <li>Read all setup and status information from the Niko Home Control Controller.
+ * <li>Execute Niko Home Control commands.
+ * <li>Listen for events from Niko Home Control.
+ * </ul>
+ *
+ * Only switch and dimmer actions are currently implemented.
+ *
+ * @author Mark Herwege - Initial Contribution
+ */
+@NonNullByDefault
+public class NikoHomeControlCommunication2 extends NikoHomeControlCommunication implements MqttMessageSubscriber {
+
+    private final Logger logger = LoggerFactory.getLogger(NikoHomeControlCommunication2.class);
+
+    private final NhcMqttConnection2 mqttConnection;
+
+    private final List<NhcProfile2> profiles = new CopyOnWriteArrayList<>();
+    private final List<NhcService2> services = new CopyOnWriteArrayList<>();
+    private final List<NhcLocation2> locations = new CopyOnWriteArrayList<>();
+
+    @Nullable
+    private volatile NhcSystemInfo2 nhcSystemInfo;
+    @Nullable
+    private volatile NhcTimeInfo2 nhcTimeInfo;
+
+    private volatile String profileUuid = "";
+
+    @Nullable
+    private volatile CompletableFuture<Boolean> communicationStarted;
+
+    private final Gson gson = new GsonBuilder().setFieldNamingPolicy(FieldNamingPolicy.UPPER_CAMEL_CASE).create();
+
+    /**
+     * Constructor for Niko Home Control communication object, manages communication with
+     * Niko Home Control II Connected Controller.
+     *
+     * @throws CertificateException when the SSL context for MQTT communication cannot be created
+     * @throws UnknownHostException when the IP address is not provided
+     *
+     */
+    public NikoHomeControlCommunication2(NhcControllerEvent handler, String clientId, String persistencePath)
+            throws CertificateException {
+        this.handler = handler;
```

Comment:
This is not really the way to do it. You should create a protected constructor in the parent class and pass the variable. That way you can make `handler` final in the parent class (assuming it's only set in constructors) and don't have to apply the `@NonNullByDefault` hack:
```suggestion
        super(handler);
```

Reply:
@Hilbrand thanks for this. Corrected in the next commit.